First implem of the lib

FlorianDoublet opened the pull request.

FlorianDoublet committed:
First dirty shot of `segmentBetween` method ( need to handle intervalExceptions )

FlorianDoublet committed:
Implement and test splitFromExceptionTimeSegments method

FlorianDoublet committed:
Split mergeTimeSegmentsForDate in mergeTimeSegments

FlorianDoublet committed:
Test method mergeTimeSegments ( and there is a bug sadly :( )

cverdier committed:
Within implem wip

FlorianDoublet committed:
Fix mergeTimeSegments

FlorianDoublet committed:
Use import scala.math.Ordering.Implicits._

cverdier committed:
factoring WIP Segments common method

cverdier committed:
tests for segments

FlorianDoublet committed:
Add method mergeSegment and mergeSegment2 ( we have to choose ) //

FlorianDoublet committed:
Add and test method `excludingSegmentFromAnother`

FlorianDoublet committed:
Fix excludingSegmentFromAnother

cverdier committed:
Clean non-segments code

FlorianDoublet committed:
Non optimized implem of exception TimeSegment in `segmentsBetween`

FlorianDoublet committed:
Merge branch 'first_implem' of github.com:Colisweb/JRubysnessHours into first_implem

FlorianDoublet committed:
Add test with a TimeSegment exception

FlorianDoublet committed:
Add more test with an exception TimeSegment

cverdier committed:
Implement within, isOpenForDurationInDate, isOpen in terms of segments

tyrcho commented on core/src/test/scala/com/colisweb/jrubysnesshours/core/SegmentsSpec.scala:
```diff
@@ -0,0 +1,384 @@
+
+  }
+
+  def aDayAt(day: String, time: String): ZonedDateTime =
```
pourquoi dans test plutot que main ces méthodes ? elles m'ont l'air bien pratiques

FlorianDoublet replied:
j'avoue elle est pratique. C'est Cyril qui a implémenté cette methode, je pense que pour le moment elle n'est utile que depuis les tests mais pourra l'être dans la partie qui communique avec JRuby :)

tyrcho commented on core/src/main/scala/com/colisweb/jrubysnesshours/core/Core.scala:
```diff
@@ -1,11 +1,103 @@
 
 object Core {
 
   case class Interval(startTime: LocalTime, endTime: LocalTime)
```
à défaut de pouvoir utiliser Refined (pas forcément pratique depuis Ruby), je propose de rajouter dans le corps de la classe un `assert(start < end)` pour "fail fast".

FlorianDoublet replied:
J'hésite, on devrait faire ce genre de fail / check ici ou alors plutot dans la partie communiquant avec JRuby ?

guizmaii replied:
La classe Interval de Threeten extra s'assure de ça : https://www.threeten.org/threeten-extra/apidocs/org.threeten.extra/org/threeten/extra/Interval.html

guizmaii replied:
Je l'ai fait dans ma PR: https://github.com/Colisweb/JRubysnessHours/pull/5/files#diff-4bd90dcefda7c996d665c7b1c807af8cR56

tyrcho commented on core/src/main/scala/com/colisweb/jrubysnesshours/core/Core.scala:
```diff
@@ -1,11 +1,103 @@
+        start,
+        end
+      )
+      .foldLeft(Duration.ZERO)(
```
j'ai l'impression qu'on gagnerait à imlémenter un Numeric[Duration] pour pouvoir écrire
```
Segments
      .segmentsBetween(planning, planningTimeZone, exceptionSegments)(
        start,
        end
      )
      .map(_.duration)
      .sum
```

avec 

```
case class TimeSegment(date: LocalDate, interval: Interval) {

    def startTime: LocalTime = interval.startTime

    def endTime: LocalTime = interval.endTime

    def duration: Duration = Duration.between(startTime, endTime)
  }
```

guizmaii replied:
Je suis très d'accord avec ta remarque Michel :)

guizmaii replied:
Cependant, `Numeric[Duration]` n'est pas implémentable car il faut implémenter une fonction `override def fromInt(x: Int): Duration`

```
.map(_.duration)
.reduce(_ plus _)
```
Fait le taf par contre.

guizmaii replied:
Fait dans ma PR

tyrcho commented on core/src/main/scala/com/colisweb/jrubysnesshours/core/Core.scala:
```diff
@@ -1,11 +1,103 @@
+      .segmentsBetween(planning, exceptionSegments)(start, end)
+      .exists(
+        segment =>
+          Duration
```
Numeric simplifierait aussi bien ici

tyrcho commented on core/src/main/scala/com/colisweb/jrubysnesshours/core/Segments.scala:
```diff
@@ -0,0 +1,215 @@
+    timeSegments: Seq[TimeSegment]
+  ): Seq[TimeSegment] = mergeTimeSegments(timeSegments.filter(_.date.isEqual(date)))
+
+  private[core] def mergeSegments(segments: Seq[TimeSegment]): List[TimeSegment] = {
```
je propose cette implémentation qui me semble plus performante (évite de parcourir la liste dans le mauvais sens : dropRight très inefficace sur une liste)
```
 private[core] def mergeSegments(
                                   segments: List[TimeSegment]
                                 ): List[TimeSegment] = {
    segments
      .sortBy(_.startTime)
      .foldRight(List.empty[TimeSegment]) {
        case (segment, h :: t) => mergeTwoSegment(segment, h) ::: t
        case (segment, Nil) => List(segment)
      }
  }
```

Note : tests OK

tyrcho commented on core/src/main/scala/com/colisweb/jrubysnesshours/core/Segments.scala:
```diff
@@ -0,0 +1,215 @@
+      }
+  }
+
+  // assuming seg1.startTime is always >= seg2.startTime ( and the same day )
```
je rajouterais un assert plutot qu'un commentaire

tyrcho commented on core/src/test/scala/com/colisweb/jrubysnesshours/core/SegmentsSpec.scala:
```diff
@@ -0,0 +1,384 @@
+  }
+
+  def aDayAt(day: String, time: String): ZonedDateTime =
+    ZonedDateTime.parse(s"${day}T$time:00.000+01:00[$zoneId]")
```
ca me parait dangereux la maniere dont on teste ici. Plus haut on définit une zone sur la TZ de Paris, et ici on a l'air de spécifier en plus GMT+1. Ca risque pas de ne plus marcher lors du passage à l'heure d'hiver ?

tyrcho commented on core/src/test/scala/com/colisweb/jrubysnesshours/core/SegmentsSpec.scala:
```diff
@@ -0,0 +1,384 @@
+class SegmentsSpec extends WordSpec with Matchers {
+
+  implicit class StringToLocalTime(str: String) {
+    def toLocalTime: LocalTime = LocalTime.parse(str)
```
si on ajoute `    def -(to: String): Interval = Interval(str.toLocalTime, to.toLocalTime)`

On peut simplifier tous les tests :
```scala
val planning: BusinessHoursByDayOfWeek = Map(
    MONDAY -> List("09:00" - "19:00"),
    TUESDAY -> List(
      "09:30" - "14:00",
      "15:00" - "19:00"
     ),
// ...
```

tyrcho committed:
simplify tests with "09:00" - "19:00" syntax

tyrcho committed:
extract common methods to SpecUtils

cverdier committed:
Refactor input data, names, exceptions type

tyrcho committed:
DSL for aDayAt and aSegment

cverdier committed:
Apply exception to segments computation

tyrcho committed:
add tests for April (based on google doc)

FlorianDoublet committed:
Fix `applyExceptionsToInterval` tests

cverdier committed:
Build Schedule cleaning intervals

cverdier committed:
Schedule contains

tyrcho committed:
reorganize segments tests for clarity

tyrcho committed:
split large test class

tyrcho committed:
first real tests with generated segments

tyrcho committed:
travis CI

tyrcho committed:
refactor PBT

cverdier committed:
Renames and formatting

tyrcho committed:
WIP merge

tyrcho committed:
Merge remote-tracking branch 'origin/first_refactoring' into tests_michem

FlorianDoublet committed:
Add JRuby object

FlorianDoublet committed:
add dateTimeIntervalsToExceptions method

guizmaii committed:
Jules' review: propose some code simplification thanks to a neat `TimeInterval` implementation :)

guizmaii committed:
Add doc

guizmaii committed:
Typo

guizmaii committed:
Clean

guizmaii committed:
Configure Travis and Codecov

guizmaii committed:
Configure Travis and Codecov

Add Travis and Codecov badges

Update scalafmt

Fix scalafmt config

scalafmt code format

Try fix CI

Try fix CI

Revert scalafmt update

guizmaii committed:
Fix code

guizmaii committed:
Add Travis and Codecov badges

guizmaii committed:
scalafmt

guizmaii committed:
Try fix CI

guizmaii committed:
Fix CI

FlorianDoublet committed:
Implement timeSegments and within methods for JRuby

guizmaii committed:
Clean

FlorianDoublet committed:
Merge pull request #6 from Colisweb/travis

Add Travis and Codecov

guizmaii committed:
Make sure that `start` is before `end` at the creation of a `TimeInterval`

guizmaii committed:
Merge branch 'first_implem' into jules_review

cverdier committed:
Testing dateTimeIntervalsToExceptions

guizmaii committed:
Clean

tyrcho committed:
restore IntervalsBetweenSpec with full tests including April

tyrcho committed:
remove duplicate test

guizmaii committed:
Review `dateTimeIntervalsToExceptions` code

guizmaii committed:
[Proposal] Alternative design for the exception subtraction algorithm

guizmaii committed:
Clean

guizmaii committed:
Fix bug

guizmaii committed:
Cyril's review: replace LocalTime.MAX by a constant which value is 23:59:00

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

cverdier committed:
Merge branch 'first_implem' into jules_review
Fix ScheduleSpec tests

cverdier committed:
Merge pull request #5 from Colisweb/jules_review

Jules' review: propose some code simplification thanks to a neat `TimeInterval` implementation :)

cverdier committed:
Merge branch 'first_implem' into proposal_alternative_design_for_exception_substraction

cverdier committed:
comment for improvement

cverdier committed:
Merge pull request #7 from Colisweb/proposal_alternative_design_for_exception_substraction

[Proposal] Alternative design for the exception subtraction algorithm

cverdier committed:
Tests for Schedule's planning

guizmaii committed:
Michel's review: The `intervalsInStartDay`, `intervalsInEndDay`, `intervalsInSameDay` methods can all be implemented with `llIntervalsInDay` and a neat `cut` mechanism

+

All the methods in `Intervals` should be in `Scheduler`.

guizmaii committed:
Merge branch 'first_implem' into michel_review

guizmaii committed:
Fix merge

guizmaii committed:
Improve code organization

guizmaii committed:
Improve code organization

guizmaii committed:
Merge branch 'michel_review' into write_gens

guizmaii committed:
Fix code for Java 8

guizmaii committed:
Fix code for Java 8

guizmaii committed:
Merge branch 'michel_review' into write_gens

guizmaii committed:
Merge branch 'write_gens' into tests_michem

guizmaii committed:
Clean

guizmaii committed:
Clean

cverdier committed:
Fix wrong period computations over multiple months

guizmaii committed:
Clean

guizmaii committed:
Merge pull request #10 from Colisweb/fix-intervals-over-many-months

Fix wrong period computations over multiple months

guizmaii committed:
Merge branch 'michel_review' into tests_michem

guizmaii committed:
Merge branch 'tests_michem' into write_gens

guizmaii committed:
Merge pull request #9 from Colisweb/write_gens

WIP: Write gens

guizmaii committed:
Clean

guizmaii committed:
Merge pull request #11 from Colisweb/write_gens

Clean

guizmaii committed:
Reorganize tests

guizmaii committed:
Merge pull request #8 from Colisweb/michel_review

Michel review

guizmaii committed:
Merge pull request #12 from Colisweb/write_gens

Reorganize tests

guizmaii committed:
Merge branch 'first_implem' into tests_michem

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Write tests for `TimeInterval#minus`

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Add more tests and fix a bug in `minus`

guizmaii committed:
Improve performances of the `TimeInterval` methods (at the price of their readability)

guizmaii committed:
Clean PR

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Add missing tests

guizmaii committed:
Add tests for `TimeInterval#contains`

guizmaii committed:
Clean

guizmaii committed:
Merge branch 'write_gens' into more_perf_minus

guizmaii committed:
Improve performances by assuming that the case where they're not connected will happen more often than the other cases

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Typo

guizmaii committed:
Merge branch 'write_gens' into more_perf_minus

guizmaii committed:
Improve performances thanks to boolean logic

FlorianDoublet committed:
Fix JRubyzSchedule

FlorianDoublet committed:
Add tests for for `rubyToDateTimeInterval` in JRubyzScheduleSpec.scala

FlorianDoublet committed:
Merge branch 'first_implem' of github.com:Colisweb/JRubysnessHours into first_implem

# Conflicts:
#	jruby/src/main/scala/com/colisweb/jrubysnesshours/jruby/JRubyzSchedule.scala

FlorianDoublet committed:
Add TODO

cverdier committed:
Next open time, few fixes of ZonedDateTimes

FlorianDoublet committed:
Add tests for rubyWeekDayToJavaWeekDay

FlorianDoublet committed:
Clean

guizmaii committed:
Merge pull request #14 from Colisweb/more_perf_minus

Improve performances of the `TimeInterval` methods (at the price of their readability)

guizmaii committed:
Merge pull request #13 from Colisweb/write_gens

Tests for TimeInterval

guizmaii committed:
Merge branch 'first_implem' into tests_michem

guizmaii committed:
Clean

FlorianDoublet committed:
Duplicate class UtilsSpec because we can't share test's code between projects :(

guizmaii committed:
Fix merge

guizmaii committed:
Merge branch 'first_implem' into tests_michem

guizmaii committed:
Merge pull request #4 from Colisweb/tests_michem

Add some tests

guizmaii committed:
Clean

FlorianDoublet committed:
change `at` method to accept a timeWithZone tuple

guizmaii committed:
Init property-based tests for `Schedule#contains`

guizmaii committed:
Clean

guizmaii committed:
Add an assert on `DateTimeInterval` creation

guizmaii committed:
WIP

cverdier committed:
Use LocalTime.MIN/MAX instead of MIDNIGHT/END_OF_DAY

guizmaii committed:
Merge pull request #17 from Colisweb/next-open-time

Implement Next open time / fix some ZonedDateTime input

guizmaii committed:
Merge branch 'first_implem' into write_gens_bis

guizmaii committed:
Merge branch 'write_gens_bis' into write_gens_ter

FlorianDoublet committed:
Fix issue

FlorianDoublet committed:
add more context to assertion failure

FlorianDoublet commented on core/src/main/scala/com/colisweb/jrubysnesshours/core/Schedule.scala:
```diff
@@ -15,76 +16,98 @@ final case class Schedule private[core] (
-    intervalsBetween(start, end).exists(_.duration >= duration)
-  }
-
-  def isOpen(instant: ZonedDateTime): Boolean = intervalsBetween(instant, instant).nonEmpty
```
Pourquoi cette méthode à disparue ? :/

FlorianDoublet committed:
Merge branch 'first_implem' into jruby_interface

# Conflicts:
#	core/src/main/scala/com/colisweb/jrubysnesshours/core/Schedule.scala

FlorianDoublet committed:
Add `nextOpentime` in JRubyzSchedule

FlorianDoublet committed:
Fix `nextOpenTimeAfter`

FlorianDoublet committed:
Fix `nextOpentime`

FlorianDoublet committed:
Fix impossible intervals creation

guizmaii committed:
Clean

guizmaii committed:
Merge pull request #19 from Colisweb/bug_fix

Bug fix on exceptions intervals

guizmaii committed:
Merge branch 'first_implem' into write_gens_bis

guizmaii committed:
Merge branch 'write_gens_bis' into write_gens_ter

guizmaii committed:
Fix bug found by the generators

guizmaii committed:
Commit missing code for previous commit

guizmaii committed:
Fix generators

guizmaii committed:
Fix tests

guizmaii committed:
Improve generators

guizmaii committed:
Clean

guizmaii committed:
Fix generators && fix a bug

guizmaii committed:
Merge pull request #18 from Colisweb/write_gens_ter

WIP

FlorianDoublet committed:
Merge branch 'first_implem' into jruby_interface

FlorianDoublet committed:
Merge pull request #16 from Colisweb/write_gens_bis

Write Schedule#contains tests

FlorianDoublet committed:
Merge branch 'first_implem' into jruby_interface

guizmaii committed:
Simplify generators

guizmaii committed:
Merge branch 'first_implem' into write_gens_ter

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

FlorianDoublet committed:
set version to 0.0.1

FlorianDoublet committed:
Setting version to 0.0.1

FlorianDoublet committed:
Fix licence version

FlorianDoublet committed:
Setting version to 0.0.2-SNAPSHOT

guizmaii committed:
Remove useless function && clean code

guizmaii committed:
Clean

guizmaii committed:
Simplify naming

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Clean

guizmaii committed:
Improve code performances

guizmaii committed:
Add method `Schedule#contains(TimeIntervalForDate)` && fix `Schedule#within` in case of empty list

FlorianDoublet committed:
Merge pull request #22 from Colisweb/remove_useless_method

Remove useless function && clean code

FlorianDoublet committed:
Remove within and contains in JRubyzSchedule

guizmaii committed:
Improve code

guizmaii committed:
Setting version to 0.0.2

guizmaii committed:
Setting version to 0.0.3-SNAPSHOT

guizmaii committed:
Setting version to 0.0.3

guizmaii committed:
Setting version to 0.0.4-SNAPSHOT

guizmaii committed:
Remove now useles `Schedule#within` method

guizmaii committed:
Improve generators and associated tests thanks to https://noelmarkham.github.io/practical-scalacheck/index.html#/16 teaching

guizmaii committed:
Merge pull request #23 from Colisweb/remove_useless_method_bis

Remove now useles `Schedule#within` method

guizmaii committed:
Clean

guizmaii committed:
Setting version to 0.0.4

guizmaii committed:
Setting version to 0.0.5-SNAPSHOT

FlorianDoublet committed:
Merge pull request #21 from Colisweb/write_gens_ter

Simplify generators

FlorianDoublet committed:
Merge branch 'first_implem' into jruby_interface

FlorianDoublet committed:
Merge branch 'jruby_interface' of github.com:Colisweb/JRubysnessHours into jruby_interface

# Conflicts:
#	core/src/main/scala/com/colisweb/jrubysnesshours/core/Schedule.scala

FlorianDoublet committed:
Merge pull request #15 from Colisweb/jruby_interface

Jruby interface
